File: web/src/lib/v2/vocabulary.ts
// v2 vocabulary: 20 math-native primitives.
// Ported from continua/vocabulary_v2.py. The Python file is the
// canonical reference for primitive definitions and the audio spec.
//
// For the human-facing Learn/Test dashboard we surface only the 16
// primitives that have standalone canonical audio. The 4 pure
// operators (ADD, MULTIPLY, NEGATE, INVERT) only sound meaningful
// when applied to arguments and live in compositional contexts.

export type WaveKind =
  | "sine"
  | "harmonic_stack"
  | "glissando"
  | "alternation"
  | "interval"
  | "spectral_sweep"
  | "sparse_burst"
  | "pulse_repetition"
  | "pitch_transform";

export type Category =
  | "quantity"
  | "operator"
  | "relation"
  | "logic"
  | "quantifier"
  | "reference"
  | "process";

export type Primitive = {
  name: string;
  category: Category;
  meaning: string;
  music: string;
  fftSignature: string;
  wave: {
    kind: WaveKind;
    frequenciesHz: number[];
    endHz?: number;
    toneDurationS?: number;
    gapS?: number;
    repeatCount?: number;
    repeatPeriodS?: number;
  };
  durationS: number;
  testable: boolean; // false for pure operators
};

const ATOMIC = 0.8;
const PROCESS = 1.2;

export const VOCABULARY_V2: Primitive[] = [
  // A. Quantity
  {
    name: "ONE",
    category: "quantity",
    meaning: "1 (the unit)",
    music: "single pitch, subitization of one object",
    fftSignature: "single peak at 440 Hz",
    wave: { kind: "sine", frequenciesHz: [440] },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "TWO",
    category: "quantity",
    meaning: "2 (octave / doubling)",
    music: "octave, universal cross-cultural identity perception",
    fftSignature: "two peaks at exact 1:2 ratio",
    wave: { kind: "harmonic_stack", frequenciesHz: [440, 880] },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "THREE",
    category: "quantity",
    meaning: "3 (triadic stack)",
    music: "root + perfect fifth + octave (1:1.5:2)",
    fftSignature: "three peaks at 1:1.5:2",
    wave: { kind: "harmonic_stack", frequenciesHz: [440, 660, 880] },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "FOUR",
    category: "quantity",
    meaning: "4 (two doublings)",
    music: "stacked octaves (1:2:4)",
    fftSignature: "three peaks at powers of 2",
    wave: { kind: "harmonic_stack", frequenciesHz: [440, 880, 1760] },
    durationS: ATOMIC,
    testable: true,
  },

  // B. Operators (pure operators have no standalone audio)
  {
    name: "ADD",
    category: "operator",
    meaning: "wave superposition: (f+g)(t) = f(t) + g(t)",
    music: "polyphony, multiple simultaneous tones",
    fftSignature: "peaks of all arguments visible together",
    wave: { kind: "harmonic_stack", frequenciesHz: [] },
    durationS: ATOMIC,
    testable: false,
  },
  {
    name: "MULTIPLY",
    category: "operator",
    meaning: "amplitude modulation: (f x g)(t) = f(t) * g(t)",
    music: "tremolo / beating, vibrating intensity",
    fftSignature: "carrier + symmetric sidebands at carrier +/- modulator",
    wave: { kind: "harmonic_stack", frequenciesHz: [] },
    durationS: ATOMIC,
    testable: false,
  },
  {
    name: "NEGATE",
    category: "operator",
    meaning: "additive inverse: NEGATE(x) + x = 0",
    music: "destructive interference, perceived in combination with argument",
    fftSignature: "magnitude identical to argument; phase inverted",
    wave: { kind: "sine", frequenciesHz: [440] },
    durationS: ATOMIC,
    testable: false,
  },
  {
    name: "INVERT",
    category: "operator",
    meaning: "multiplicative inverse: f to 440^2/f",
    music: "pitch mirror around fundamental",
    fftSignature: "peak(s) at 440^2/f for each argument frequency",
    wave: { kind: "sine", frequenciesHz: [440] },
    durationS: ATOMIC,
    testable: false,
  },

  // C. Relations
  {
    name: "EQUAL",
    category: "relation",
    meaning: "identity",
    music: "unison, universal fusion of identical tones",
    fftSignature: "single peak (doubled amplitude permitted)",
    wave: { kind: "sine", frequenciesHz: [440] },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "GREATER",
    category: "relation",
    meaning: "positive comparison",
    music: "ascending pitch contour, universal 'up / more'",
    fftSignature: "time-varying peak sliding upward",
    wave: { kind: "glissando", frequenciesHz: [440], endHz: 880 },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "LESSER",
    category: "relation",
    meaning: "negative comparison",
    music: "descending pitch contour, universal 'down / less'",
    fftSignature: "time-varying peak sliding downward",
    wave: { kind: "glissando", frequenciesHz: [880], endHz: 440 },
    durationS: ATOMIC,
    testable: true,
  },

  // D. Logic
  {
    name: "AND",
    category: "logic",
    meaning: "conjunction (perfect fifth bind)",
    music: "perfect fifth (3:2), second-most-consonant interval",
    fftSignature: "two peaks at exact 3:2 ratio",
    wave: { kind: "interval", frequenciesHz: [440, 660] },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "OR",
    category: "logic",
    meaning: "disjunction (rapid alternation)",
    music: "stream segregation, distinct events in time",
    fftSignature: "peaks at 440 and 660 Hz in alternating time windows",
    wave: { kind: "alternation", frequenciesHz: [440, 660], toneDurationS: 0.2, gapS: 0 },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "NOT",
    category: "logic",
    meaning: "logical negation (dissonance)",
    music: "minor second (16:15), most dissonant interval; universal 'wrong / tension'",
    fftSignature: "two peaks at 16:15 ratio; audible beating",
    wave: { kind: "interval", frequenciesHz: [440, 466.16] },
    durationS: ATOMIC,
    testable: true,
  },

  // E. Quantifiers
  {
    name: "ALL",
    category: "quantifier",
    meaning: "universal quantifier",
    music: "five-octave glissando, 'every pitch / everything'",
    fftSignature: "time-varying peak traversing full spectrum",
    wave: { kind: "spectral_sweep", frequenciesHz: [110], endHz: 3520 },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "SOME",
    category: "quantifier",
    meaning: "existential quantifier",
    music: "scattered discrete events, 'some specific things'",
    fftSignature: "sparse peaks at non-consecutive harmonics",
    wave: { kind: "sparse_burst", frequenciesHz: [880, 1320, 2200], toneDurationS: 0.1, gapS: 0.15 },
    durationS: ATOMIC,
    testable: true,
  },

  // F. Reference
  {
    name: "SELF",
    category: "reference",
    meaning: "speaker / message frame of origin",
    music: "low-frequency centered tone, 'grounded, here, body'",
    fftSignature: "peak at 110 Hz",
    wave: { kind: "sine", frequenciesHz: [110] },
    durationS: ATOMIC,
    testable: true,
  },
  {
    name: "TARGET",
    category: "reference",
    meaning: "referent / object of the message",
    music: "high-frequency tone, 'over there, the thing'",
    fftSignature: "peak at 1760 Hz",
    wave: { kind: "sine", frequenciesHz: [1760] },
    durationS: ATOMIC,
    testable: true,
  },

  // G. Process
  {
    name: "BECOMES",
    category: "process",
    meaning: "continuous transformation: f: t to x(t)",
    music: "portamento, vocal pitch glide; universal 'becoming'",
    fftSignature: "spectral peak sliding continuously between source and target",
    wave: { kind: "pitch_transform", frequenciesHz: [440], endHz: 880 },
    durationS: PROCESS,
    testable: true,
  },
  {
    name: "REPEATS",
    category: "process",
    meaning: "periodic function: x(t+T) = x(t)",
    music: "regular pulse / meter, beat induction universal even in infants",
    fftSignature: "repeated spectral signature; autocorrelation peak at period",
    wave: { kind: "pulse_repetition", frequenciesHz: [440], repeatCount: 4, repeatPeriodS: 0.3 },
    durationS: PROCESS,
    testable: true,
  },
];

export const TESTABLE_V2 = VOCABULARY_V2.filter((p) => p.testable);
export const V2_CHANCE_RATE = 1.0 / TESTABLE_V2.length;

File: continua/v3/vocabulary.py
"""v3 vocabulary — 20 primitives, each fully specified across 4 feature axes.

Per docs/vocabulary_v3_spec.md:
  - frequency family : category-specific root, families do not overlap
  - timbre           : spectral envelope (sine | square | sawtooth | triangle | formant | noise_tone | envelope_shaped)
  - spatial position : "center" | "right" | "left" | "diffuse"
  - temporal envelope: steady | sustained | rising | falling | glide | alternating | dissonant | sweep | scattered | throb | shimmer | superposition | am | sharp | reflection | repeated

Schema for messages stays the same JSON shape as v2 (subject/relation/object
phrases, metadata, implies_next) but version is "3.0".
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Literal, Optional, Tuple


# ---------------------------------------------------------------------------
# Category roots (no overlap between categories)
# ---------------------------------------------------------------------------

FAMILY_ROOT: Dict[str, float] = {
    "quantity":   220.0,   # A3
    "logic":      165.0,   # E3
    "relation":   330.0,   # E4
    "operator":   660.0,   # E5
    "quantifier": 110.0,   # A2 root, broadband
    "reference":  None,    # spatial + sub-bass/high anchor, not a single root
    "process":    None,    # defined by temporal envelope
}

Category = Literal["quantity", "logic", "relation", "operator",
                    "quantifier", "reference", "process"]

Timbre = Literal["sine", "square", "sawtooth", "triangle", "formant",
                  "noise_tone", "envelope_shaped"]

Spatial = Literal["center", "right", "left", "diffuse"]

TemporalShape = Literal[
    "steady", "sustained", "rising", "falling", "glide",
    "alternating", "dissonant", "sweep", "scattered",
    "throb", "shimmer", "superposition", "am", "sharp",
    "reflection", "repeated",
]

Grounding = Literal["strong", "moderate", "reach"]


# ---------------------------------------------------------------------------
# Primitive dataclass — 4-feature specification
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class Primitive:
    name: str
    category: Category
    math_meaning: str

    # Axis 1: frequency family
    family_root_hz: Optional[float]
    partial_ratios: Tuple[float, ...]  # multiples of family root (e.g. (1.0, 2.0) for octave)

    # Axis 2: timbre
    timbre: Timbre

    # Axis 3: spatial position
    spatial: Spatial

    # Axis 4: temporal envelope
    temporal: TemporalShape

    # Standard primitive metadata
    duration_s: float
    grounding: Grounding
    takes_args: bool = False
    accepts_modifier: bool = False
    music_basis: str = ""

    def absolute_frequencies(self) -> Tuple[float, ...]:
        """Compute concrete Hz values from family root × partial ratios."""
        if self.family_root_hz is None:
            return ()
        return tuple(self.family_root_hz * r for r in self.partial_ratios)


ATOMIC_DUR = 0.8
PROCESS_DUR = 1.2


# ---------------------------------------------------------------------------
# The 20 primitives
# ---------------------------------------------------------------------------

VOCABULARY_V3: Tuple[Primitive, ...] = (
    # --- Quantity (4) — family 220 Hz, sine timbre, center, steady ---
    Primitive(
        name="ONE", category="quantity", math_meaning="1 (the unit)",
        family_root_hz=220.0, partial_ratios=(1.0,),
        timbre="sine", spatial="center", temporal="steady",
        duration_s=ATOMIC_DUR, grounding="strong",
        music_basis="single sine tone — subitization of one object",
    ),
    Primitive(
        name="TWO", category="quantity", math_meaning="2 (octave / doubling)",
        family_root_hz=220.0, partial_ratios=(1.0, 2.0),
        timbre="sine", spatial="center", temporal="steady",
        duration_s=ATOMIC_DUR, grounding="strong",
        music_basis="octave (2:1) — universal cross-cultural identity",
    ),
    Primitive(
        name="THREE", category="quantity", math_meaning="3 (1:2:3 triadic stack)",
        family_root_hz=220.0, partial_ratios=(1.0, 2.0, 3.0),
        timbre="sine", spatial="center", temporal="steady",
        duration_s=ATOMIC_DUR, grounding="strong",
        music_basis="harmonic series 1:2:3 — fundamental + octave + perfect fifth above octave",
    ),
    Primitive(
        name="FOUR", category="quantity", math_meaning="4 (two doublings)",
        family_root_hz=220.0, partial_ratios=(1.0, 2.0, 4.0),
        timbre="sine", spatial="center", temporal="steady",
        duration_s=ATOMIC_DUR, grounding="strong",
        music_basis="1:2:4 stacked octaves",
    ),

    # --- Logic (3) — family 165 Hz, triangle timbre, center, sustained ---
    Primitive(
        name="AND", category="logic", math_meaning="conjunction (perfect fifth bind)",
        family_root_hz=165.0, partial_ratios=(1.0, 1.5),
        timbre="triangle", spatial="center", temporal="sustained",
        duration_s=ATOMIC_DUR, grounding="strong", takes_args=True,
        music_basis="perfect fifth (3:2) in the logic family",
    ),
    Primitive(
        name="OR", category="logic", math_meaning="disjunction (alternation)",
        family_root_hz=165.0, partial_ratios=(1.0, 1.5),
        timbre="triangle", spatial="center", temporal="alternating",
        duration_s=ATOMIC_DUR, grounding="strong", takes_args=True,
        music_basis="rapid alternation between two states",
    ),
    Primitive(
        name="NOT", category="logic", math_meaning="logical negation (dissonance)",
        family_root_hz=165.0, partial_ratios=(1.0, 16.0 / 15.0),
        timbre="triangle", spatial="center", temporal="dissonant",
        duration_s=ATOMIC_DUR, grounding="moderate", takes_args=True,
        music_basis="minor second (16:15) — universal dissonance / 'wrong'",
    ),

    # --- Relations (4) — family 330 Hz, sawtooth timbre, center, motion ---
    Primitive(
        name="EQUAL", category="relation", math_meaning="identity",
        family_root_hz=330.0, partial_ratios=(1.0,),
        timbre="sawtooth", spatial="center", temporal="sustained",
        duration_s=ATOMIC_DUR, grounding="strong", accepts_modifier=True,
        music_basis="unison sustained at 330 — identity tone",
    ),
    Primitive(
        name="GREATER", category="relation", math_meaning="positive comparison",
        family_root_hz=330.0, partial_ratios=(1.0, 2.0),  # endpoints of glide
        timbre="sawtooth", spatial="center", temporal="rising",
        duration_s=ATOMIC_DUR, grounding="strong", accepts_modifier=True,
        music_basis="rising glissando — universal 'up / more'",
    ),
    Primitive(
        name="LESSER", category="relation", math_meaning="negative comparison",
        family_root_hz=330.0, partial_ratios=(2.0, 1.0),
        timbre="sawtooth", spatial="center", temporal="falling",
        duration_s=ATOMIC_DUR, grounding="strong", accepts_modifier=True,
        music_basis="falling glissando — universal 'down / less'",
    ),
    Primitive(
        name="BECOMES", category="relation", math_meaning="continuous transformation",
        family_root_hz=330.0, partial_ratios=(1.0, 2.0),
        timbre="sawtooth", spatial="center", temporal="glide",
        duration_s=PROCESS_DUR, grounding="strong", accepts_modifier=True,
        music_basis="portamento — vocal pitch glide; universal 'becoming'",
    ),

    # --- Operators (4) — family 660 Hz, square timbre, center, sharp ---
    Primitive(
        name="ADD", category="operator", math_meaning="wave superposition: f+g",
        family_root_hz=660.0, partial_ratios=(1.0,),
        timbre="square", spatial="center", temporal="superposition",
        duration_s=ATOMIC_DUR, grounding="strong", takes_args=True,
        music_basis="polyphony — multiple simultaneous tones",
    ),
    Primitive(
        name="MULTIPLY", category="operator", math_meaning="amplitude modulation: f*g",
        family_root_hz=660.0, partial_ratios=(1.0,),
        timbre="square", spatial="center", temporal="am",
        duration_s=ATOMIC_DUR, grounding="reach", takes_args=True,
        music_basis="tremolo / sideband formation",
    ),
    Primitive(
        name="NEGATE", category="operator", math_meaning="additive inverse",
        family_root_hz=660.0, partial_ratios=(1.0,),
        timbre="square", spatial="center", temporal="sharp",
        duration_s=ATOMIC_DUR, grounding="reach", takes_args=True,
        music_basis="phase-inverted argument; cancels argument when summed",
    ),
    Primitive(
        name="INVERT", category="operator", math_meaning="multiplicative inverse (frequency reflection)",
        family_root_hz=660.0, partial_ratios=(1.0,),
        timbre="square", spatial="center", temporal="reflection",
        duration_s=ATOMIC_DUR, grounding="moderate", takes_args=True,
        music_basis="frequency mirror around 660 Hz",
    ),

    # --- Quantifiers (2) — broadband, noise+tone, diffuse, sweep/scattered ---
    Primitive(
        name="ALL", category="quantifier", math_meaning="universal quantifier",
        family_root_hz=110.0, partial_ratios=(1.0, 32.0),  # 110 → 3520 Hz
        timbre="noise_tone", spatial="diffuse", temporal="sweep",
        duration_s=ATOMIC_DUR, grounding="moderate",
        music_basis="exponential pitch sweep across 5 octaves — 'every pitch'",
    ),
    Primitive(
        name="SOME", category="quantifier", math_meaning="existential quantifier",
        family_root_hz=110.0, partial_ratios=(2.0, 4.0, 8.0),  # A3, A4, A5 scattered
        timbre="noise_tone", spatial="diffuse", temporal="scattered",
        duration_s=ATOMIC_DUR, grounding="reach",
        music_basis="discrete sparse events — 'some specific things'",
    ),

    # --- References (2) — spatial + sub-bass/high anchor, formant timbre ---
    Primitive(
        name="SELF", category="reference", math_meaning="speaker / message frame of origin",
        family_root_hz=55.0, partial_ratios=(1.0,),  # 55 Hz sub-bass
        timbre="formant", spatial="center", temporal="throb",
        duration_s=ATOMIC_DUR, grounding="strong",
        music_basis="centered sub-bass with slow ~4 Hz amplitude throb — body-felt 'I'",
    ),
    Primitive(
        name="TARGET", category="reference", math_meaning="referent / addressed entity",
        family_root_hz=1760.0, partial_ratios=(1.0,),  # 1760 Hz bright
        timbre="formant", spatial="right", temporal="shimmer",
        duration_s=ATOMIC_DUR, grounding="strong",
        music_basis="right-hemifield bright tone with ~7 Hz shimmer — 'over there, you'",
    ),

    # --- Process (1) — defined by temporal envelope ---
    Primitive(
        name="REPEATS", category="process", math_meaning="periodic function x(t+T)=x(t)",
        family_root_hz=440.0, partial_ratios=(1.0,),  # carrier tone for the pulses
        timbre="envelope_shaped", spatial="center", temporal="repeated",
        duration_s=PROCESS_DUR, grounding="strong",
        music_basis="regular pulse train at 0.3s period — beat induction is universal",
    ),
)


PRIMITIVES_BY_NAME: Dict[str, Primitive] = {p.name: p for p in VOCABULARY_V3}

OPERATORS: Tuple[str, ...] = tuple(p.name for p in VOCABULARY_V3 if p.takes_args)
RELATIONS: Tuple[str, ...] = tuple(p.name for p in VOCABULARY_V3 if p.accepts_modifier)
TIMBRES_BY_CATEGORY: Dict[str, str] = {
    p.category: p.timbre for p in VOCABULARY_V3
}
PRIMITIVES_BY_CATEGORY: Dict[str, Tuple[Primitive, ...]] = {}
for p in VOCABULARY_V3:
    PRIMITIVES_BY_CATEGORY.setdefault(p.category, ())
    PRIMITIVES_BY_CATEGORY[p.category] = PRIMITIVES_BY_CATEGORY[p.category] + (p,)


# ---------------------------------------------------------------------------
# Message validation (v3.0 schema — same shape as v2.1 but version "3.0")
# ---------------------------------------------------------------------------

SUPPORTED_VERSIONS = ("3.0",)
PHRASE_ARG_OPERATORS = ("AND", "OR")


@dataclass
class ValidationError(Exception):
    path: str
    reason: str
    def __str__(self) -> str:
        return f"at {self.path}: {self.reason}"


def _is_phrase(node: dict) -> bool:
    return all(k in node for k in ("subject", "relation", "object"))


def _validate_primitive_node(node: dict, path: str) -> None:
    if "primitive" not in node:
        raise ValidationError(path, "missing 'primitive'")
    name = node["primitive"]
    if name not in PRIMITIVES_BY_NAME:
        raise ValidationError(path, f"unknown primitive '{name}'")
    prim = PRIMITIVES_BY_NAME[name]
    if "args" in node:
        if not prim.takes_args:
            raise ValidationError(path, f"'{name}' does not take args")
        args = node["args"]
        if not isinstance(args, list) or not args:
            raise ValidationError(path, f"'{name}' args must be non-empty list")
        for i, a in enumerate(args):
            ap = f"{path}.args[{i}]"
            if name in PHRASE_ARG_OPERATORS and _is_phrase(a):
                _validate_phrase(a, ap)
            else:
                _validate_primitive_node(a, ap)
    if "modifier" in node:
        if not prim.accepts_modifier:
            raise ValidationError(path, f"'{name}' does not accept modifier")
        if node["modifier"] not in PRIMITIVES_BY_NAME:
            raise ValidationError(path, f"unknown modifier '{node['modifier']}'")


def _validate_phrase(phrase: dict, base: str) -> None:
    for slot in ("subject", "relation", "object"):
        if slot not in phrase:
            raise ValidationError(base, f"missing '{slot}'")
        _validate_primitive_node(phrase[slot], f"{base}.{slot}")
    if "implies_next" in phrase and not isinstance(phrase["implies_next"], bool):
        raise ValidationError(base, "'implies_next' must be a boolean")


def validate_message(msg: dict) -> None:
    if msg.get("type") != "continua_v3_message":
        raise ValidationError("$", "type must be 'continua_v3_message'")
    if msg.get("version") not in SUPPORTED_VERSIONS:
        raise ValidationError("$", f"version must be one of {SUPPORTED_VERSIONS}")
    phrases = msg.get("phrases")
    if not isinstance(phrases, list) or not phrases:
        raise ValidationError("$.phrases", "must be non-empty list")
    for i, phrase in enumerate(phrases):
        _validate_phrase(phrase, f"$.phrases[{i}]")
    if phrases[-1].get("implies_next"):
        raise ValidationError(
            f"$.phrases[{len(phrases)-1}]",
            "last phrase cannot set 'implies_next'",
        )


def chance_rate() -> float:
    return 1.0 / len(VOCABULARY_V3)
